libs/vkd3d/resource.c

```diff
@@ -209,15 +209,40 @@ static unsigned int max_miplevel_count(const D3D12_RESOURCE_DESC1 *desc)
     return vkd3d_log2i(size) + 1;
 }
 
-static const struct vkd3d_format_compatibility_list *vkd3d_get_format_compatibility_list(
-        const struct d3d12_device *device, DXGI_FORMAT dxgi_format)
+static bool vkd3d_get_format_compatibility_list(const struct d3d12_device *device,
+        const D3D12_RESOURCE_DESC1 *desc, struct vkd3d_format_compatibility_list *out_list)
 {
-    const struct vkd3d_format_compatibility_list *list = &device->format_compatibility_lists[dxgi_format];
+    static const VkFormat r32_uav_formats[] = { VK_FORMAT_R32_UINT, VK_FORMAT_R32_SINT, VK_FORMAT_R32_SFLOAT };
+    const struct vkd3d_format *format = vkd3d_get_format(device, desc->Format, false);
+    struct vkd3d_format_compatibility_list list;
+    unsigned int i;
 
-    if (list->format_count < 2)
-        return NULL;
+    memset(&list, 0, sizeof(list));
+
+    if (desc->Format < device->format_compatibility_list_count)
+        list = device->format_compatibility_lists[desc->Format];
+
+    if (desc->Flags & D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS)
+    {
+        const struct vkd3d_format *uint_format = vkd3d_get_format(device, list.uint_format, false);
+
+        /* Format used for ClearUnorderedAccessViewUint */
+        if (uint_format)
+            vkd3d_format_compatibility_list_add_format(&list, uint_format->vk_format);
+
+        /* Legacy D3D11 compatibility rule that allows typed UAV loads on FL11.0 hardware */
+        if (format->type == VKD3D_FORMAT_TYPE_TYPELESS && format->byte_count == 4)
```

Comment:
I'm not sure this is correct. In the test I observed that it was possible to cast RGBA8_UNORM to R32UI UAV, but not R32_FLOAT. Is the D3D11 legacy rule only for TYPELESS? Might be a bug in the validation layer fwiw.

Reply:
I'm not sure what the exact rules are here. Although I suppose it makes sense to allow non-typeless since there's no R11G11B10 typeless format but you can still use it for UAVs.

Reply:
I'll add some tests for that case.